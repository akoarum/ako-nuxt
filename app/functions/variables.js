export const VALIDATES = {
  EMPTY: {
    message: '入力してください。'
  },
  NUMBER: {
    regex: /^[0-9]+?$/,
    message: '数字で入力してください。'
  },
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^[0-9]{2,4}-?[0-9]{2,4}-?[0-9]{4}$/,
    message: '電話番号の形式になっていません。'
  },
  URL: {
    regex: /^http(s):\/\/.+?\.+?$/,
    message: 'URLの形式になっていません。'
  }
}
