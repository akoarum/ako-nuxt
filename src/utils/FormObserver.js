export default class FormValidator {
  constructor(names = []) {
    this._items = this._createItems(names)
  }

  /**
   * @return {Boolean}
   */
  get btnDisabled() {
    for (const name in this._items) {
      if (this._items[name].error || !this._items[name].dirty) return true
    }
    return false
  }

  /**
   * itemオブジェクトの生成
   * @param {Array} names name配列
   */
  _createItems(names) {
    const items = {}

    names.forEach(item => {
      items[item] = {
        error: false,
        dirty: false
      }
    })

    return items
  }

  /**
   * itemの更新
   * @param {String}  name  name値
   * @param {String}  prop  更新するプロパティ名
   * @param {Boolean} value 値
   */
  updateItem(name, prop, value) {
    this._items[name][prop] = value
  }
}
