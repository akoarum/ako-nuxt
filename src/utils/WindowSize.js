export default class WindowSize {
  constructor(delay = 80) {
    this._width = 0
    this._height = 0
    this._delay = delay
    this._cullTime = null
    this._initialized = false
    this._handler = this._handleResize.bind(this)
  }

  get width() {
    return this._width
  }

  get height() {
    return this._height
  }

  get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  initialize() {
    if (this._initialized) return this
    this._updateState()
    window.addEventListener('resize', this._handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  _updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  _handleResize() {
    clearTimeout(this._cullTime)
    this._cullTime = setTimeout(() => {
      this._updateState()
    }, this._delay)
  }

  /**
   * 解放する
   */
  destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this._handler)
    this._initialized = false
    return this
  }
}
