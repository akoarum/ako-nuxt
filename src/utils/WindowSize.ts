export default class WindowSize {
  private _width: number = 0
  private _height: number = 0
  private delay: number = 80
  private cullTime = null
  private _initialized: boolean = false
  private handler = this.handleResize.bind(this)

  constructor(delay: number = 80) {
    this.delay = delay
  }

  public get width() {
    return this._width
  }

  public get height() {
    return this._height
  }

  public get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  public initialize() {
    if (this._initialized) return this
    this.updateState()
    window.addEventListener('resize', this.handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  public updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  private handleResize() {
    clearTimeout(this.cullTime)
    this.cullTime = setTimeout(() => {
      this.updateState()
    }, this.delay)
  }

  /**
   * 解放する
   */
  public destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this.handler)
    this._initialized = false
    return this
  }
}
