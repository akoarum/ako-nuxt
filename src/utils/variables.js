export const VALIDATES = {
  required: '入力してください。',
  selected: '選択してください。',
  file: 'ファイルをアップロードしてください。',
  tel: {
    regex: /^\d{2,4}-?\d{4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
