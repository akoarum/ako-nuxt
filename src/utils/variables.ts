import { ValidatePattern } from '~/types/models'

export const VALIDATE_PATTERNS: { [key: string]: ValidatePattern } = {
  required: {
    message: '入力してください。'
  },
  selected: {
    message: '選択してください。'
  },
  tel: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
