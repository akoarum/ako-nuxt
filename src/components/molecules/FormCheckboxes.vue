<template>
  <div class="formCheckboxes">
    <ul class="formCheckboxes__list">
      <li
        v-for="select in selects"
        :key="select.id"
        class="formCheckboxes__item"
      >
        <VCheckbox
          v-model="model"
          v-bind="$attrs"
          :name="name"
          :value="select.value"
          :aria-invalid="String(!!error)"
          :aria-describedby="error ? `error-${name}` : false"
          @change.once="updateDirty"
        >
          {{ select.label }}
        </VCheckbox>
      </li>
    </ul>
    <VTexts
      v-if="error"
      :id="`error-${name}`"
      role="alert"
      aria-live="assertive"
      small
      caution
      class="formCheckboxes__error"
    >
      <span>{{ error }}</span>
    </VTexts>
  </div>
</template>

<script>
import { VALIDATES } from '~/utils/variables'
import VCheckbox from '~/components/atoms/VCheckbox'
import VTexts from '~/components/atoms/VTexts'

export default {
  components: { VCheckbox, VTexts },
  inheritAttrs: false,
  model: {
    event: 'change'
  },
  props: {
    name: { type: String, required: true },
    selects: { type: Array, required: true },
    value: { type: Array },
    required: { type: Boolean },
    min: { type: Number },
    max: { type: Number }
  },
  data() {
    return {
      error: '',
      isDirty: !!this.value.length || !this.required
    }
  },
  computed: {
    model: {
      set(value) {
        this.$emit('change', value)
        this.error = this.validate(value)
      },
      get() {
        return this.value
      }
    }
  },
  watch: {
    error(value) {
      this.$emit('error', { name: this.name, error: !!value })
    },
    isDirty() {
      this.$emit('dirty', this.name)
    }
  },
  methods: {
    validate(value) {
      if (!this.required) return ''
      if (!this.isDirty) return ''
      if (!value.length) return VALIDATES.selected
      if (this.min && value.length < this.min) {
        return `${this.min}個以上選択してください。`
      }
      if (this.max && value.length > this.max) {
        return `${this.max}個以内で選択してください。`
      }
    },
    updateDirty() {
      this.isDirty = true
    }
  }
}
</script>

<style lang="scss" scoped>
.formCheckboxes__item + .formCheckboxes__item {
  margin-top: 12px;
}
.formCheckboxes__error {
  @include media-min {
    margin-top: 6px;
  }
  @include media-max {
    margin-top: 8px;
  }
}
</style>
