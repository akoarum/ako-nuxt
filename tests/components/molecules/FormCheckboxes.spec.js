import { mount } from '@vue/test-utils'
import { VALIDATES } from '~/utils/variables'
import FormCheckboxes from '~/components/molecules/FormCheckboxes'

describe('FormCheckboxes', () => {
  let wrapper
  const selects = [
    { id: 1, label: '1', value: '1' },
    { id: 2, label: '2', value: '2' },
    { id: 3, label: '3', value: '3' }
  ]

  beforeEach(() => {
    wrapper = mount(FormCheckboxes, {
      propsData: {
        name: 'test',
        value: [],
        selects,
        required: true
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('validate', () => {
      it('requiredでない場合は空文字列を返す', () => {
        wrapper.setProps({ required: false })
        expect(wrapper.vm.validate([])).toBe('')
      })

      it('isDirtyでない場合も空文字列を返す', () => {
        wrapper.setData({ isDirty: false })
        expect(wrapper.vm.validate([])).toBe('')
      })

      it('requiredでisDirty、valueが空配列の場合はselectedエラーを返す', () => {
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate([])).toBe(VALIDATES.selected)
      })

      it('minが設定されていて、valueの数が足りない場合はエラーを返す', () => {
        wrapper.setProps({ min: 2 })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate(['1'])).toBe('2個以上選択してください。')
      })

      it('maxが設定されていて、valueの数が超えている場合はエラーを返す', () => {
        wrapper.setProps({ max: 2 })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate(['1', '2', '3'])).toBe('2個以内で選択してください。')
      })
    })

    describe('updateDirty', () => {
      it('isDirtyをtrueにする', () => {
        wrapper.vm.updateDirty()
        expect(wrapper.vm.isDirty).toBe(true)
      })
    })
  })
})
