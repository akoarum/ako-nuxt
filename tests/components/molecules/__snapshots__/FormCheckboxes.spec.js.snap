// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxes initialized snapshot 1`] = `
<div
  class="formCheckboxes"
>
  <ul
    class="formCheckboxes__list"
  >
    <li
      class="formCheckboxes__item"
    >
      <label
        class="vCheckbox"
      >
        <input
          aria-invalid="false"
          class="vCheckbox__input"
          name="test"
          type="checkbox"
          value="1"
        />
         
        <span
          class="vCheckbox__label"
        >
          
        1
      
        </span>
      </label>
    </li>
    <li
      class="formCheckboxes__item"
    >
      <label
        class="vCheckbox"
      >
        <input
          aria-invalid="false"
          class="vCheckbox__input"
          name="test"
          type="checkbox"
          value="2"
        />
         
        <span
          class="vCheckbox__label"
        >
          
        2
      
        </span>
      </label>
    </li>
    <li
      class="formCheckboxes__item"
    >
      <label
        class="vCheckbox"
      >
        <input
          aria-invalid="false"
          class="vCheckbox__input"
          name="test"
          type="checkbox"
          value="3"
        />
         
        <span
          class="vCheckbox__label"
        >
          
        3
      
        </span>
      </label>
    </li>
  </ul>
   
  <!---->
</div>
`;
